src/tsolvers/axdiffsolver/ArraySolver.cc

```diff
@@ -432,104 +424,89 @@ PTRef ArraySolver::computeExtensionalityClause(NodeRef n1, NodeRef n2) {
-        ERef root = egraph.getRoot(index);
+        ERef root = getRoot(getIndexFromSelect(select));
         indicesToSelects[root].push(select);
     }
-    for (auto const & entry : indicesToSelects) {
-        ERef index = entry.first;
-        auto const & selects = entry.second;
+    std::vector<LemmaConditions> lemmas;
+    for (auto const & [index, selects] : indicesToSelects) {
         if (selects.size() < 2) { continue; }
         // TODO: Figure out better way how to compute all candidates for lemmas
-        for (auto firstIt = selects.begin(); firstIt != selects.end(); ++firstIt) {
-            ERef first = *firstIt;
+        for (auto first : selects) {
             ERef firstRoot = getRoot(first);
-            for (auto secondIt = selects.begin(); secondIt != firstIt; ++secondIt) {
+            for (auto secondIt = selects.begin(); *secondIt != first; ++secondIt) {
                 ERef second = *secondIt;
                 if (firstRoot == getRoot(second)) { continue; } // The selects are already the same, no lemma needed
                 NodeRef arrayFirst = getNodeRef(getRoot(getArrayFromSelect(first)));
                 NodeRef arraySecond = getNodeRef(getRoot(getArrayFromSelect(second)));
                 if (arrayFirst == arraySecond or getIndexedRepresentative(arrayFirst, index) == getIndexedRepresentative(arraySecond, index)) {
-                    LemmaConditions lemmaConditions;
-                    PTRef equalityOfSelects = getEquality(first, second);
-                    lemmaConditions.equality = equalityOfSelects;
+                    std::unordered_set<PTRef, PTRefHash> undecidedEqualities;
+                    PTRef equalityOfSelects = getEquality(first, second, logic);
                     if (not isFalsified(equalityOfSelects)) {
                         assert(not isSatisfied(equalityOfSelects));
-                        lemmaConditions.undecidedEqualities.insert(equalityOfSelects);
+                        undecidedEqualities.insert(equalityOfSelects);
                     }
                     auto storeIndices = Traversal(*this).computeStoreIndices(arrayFirst, arraySecond, index);
                     for (ERef storeIndex : storeIndices) {
                         assert(storeIndex != index);
-                        PTRef equalityOfIndices = getEquality(index, storeIndex);
+                        PTRef equalityOfIndices = getEquality(index, storeIndex, logic);
                         if (not isFalsified(equalityOfIndices)) {
                             assert(not isSatisfied(equalityOfIndices));
-                            lemmaConditions.undecidedEqualities.insert(equalityOfIndices);
+                            undecidedEqualities.insert(equalityOfIndices);
                         }
                     }
-                    this->lemmas.push_back(std::move(lemmaConditions));
+                    lemmas.emplace_back(LemmaConditions{equalityOfSelects, std::move(undecidedEqualities)});
```

Comment:
```suggestion
                    lemmas.emplace_back(equalityOfSelects, std::move(undecidedEqualities));
```

Reply:
OK, it looks like this does not work.
Then probably revert this and I'll accept your previous version.

Reply:
Damn, that didn't seem to work...